dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/acl/DefaultTrackerOwnershipManager.java

```diff
@@ -104,26 +113,28 @@ public DefaultTrackerOwnershipManager(
   /** Cache for storing recent temporary ownership checks */
   private final Cache<Boolean> tempOwnerCache;
 
-  // -------------------------------------------------------------------------
-  // Implementation
-  // -------------------------------------------------------------------------
-
   @Override
   @Transactional
-  public void transferOwnership(
-      TrackedEntity trackedEntity, Program program, OrganisationUnit orgUnit)
-      throws ForbiddenException {
-    if (trackedEntity == null || program == null || orgUnit == null) {
-      return;
+  // TODO(tracker) This method should accept a tracked entity UID instead. The problem is, we can't
+  // use the TrackedEntityService as it introduces a cyclic dependency. That's because the ownership
+  // manager is used to filter TEs after hitting the database. As soon as we move those filters into
+  // the store to fix the pagination issue, we should be able to use the TrackedEntityService here,
+  // so we can run all validations in this service instead of the controller.
+  public void transferOwnership(TrackedEntity trackedEntity, UID programUid, UID orgUnitUid)
+      throws ForbiddenException, BadRequestException {
+    Program program = trackerProgramService.getTrackerProgram(programUid);
+    OrganisationUnit orgUnit = organisationUnitService.getOrganisationUnit(orgUnitUid.getValue());
+
+    if (orgUnit == null) {
+      throw new ForbiddenException(
+          "Tracked entity not transferred. Org unit supplied does not exist.");
     }
 
-    UserDetails currentUser = CurrentUserUtil.getCurrentUserDetails();
-
-    if (hasAccess(currentUser, trackedEntity, program)) {
+    if (hasAccess(getCurrentUserDetails(), trackedEntity, program)) {
```

Comment:
Perhaps we can flip some of these if conditions to have easy to read flow (avoids chained or inner ifs)

Like

if (orgUnit == null)
throw

if(!hasAccess....)
throw

if (!programService.hasOrgUnit(program, orgUnit))
throw

if (any other validation failure)
throw

....normal success flow...